src/Repeater/DefaultRepeaterServer.ts

```diff
@@ -0,0 +1,173 @@
+import { logger } from '../Utils';
+import {
+  RepeaterServer,
+  RepeaterServerEventHandler,
+  RepeaterServerDeployedEvent,
+  RepeaterServerEvents,
+  RepeaterServerReconnectionFailedEvent,
+  RepeaterServerEventHandlers
+} from './RepeaterServer';
+import { inject, injectable } from 'tsyringe';
+import io, { Socket } from 'socket.io-client';
+import parser from 'socket.io-msgpack-parser';
+import { once } from 'events';
+import Timer = NodeJS.Timer;
+
+export interface DefaultRepeaterServerOptions {
+  readonly uri: string;
+  readonly token: string;
+}
+
+export const DefaultRepeaterServerOptions: unique symbol = Symbol(
+  'DefaultRepeaterServerOptions'
+);
+
+@injectable()
+export class DefaultRepeaterServer implements RepeaterServer {
+  private latestReconnectionError?: Error;
+  private readonly DEFAULT_RECONNECT_TIMES = 20;
+  private socket?: Socket;
+  private timer?: Timer;
+
+  constructor(
+    @inject(DefaultRepeaterServerOptions)
+    private readonly options: DefaultRepeaterServerOptions
+  ) {}
+
+  public disconnect() {
+    this.clearPingTimer();
+
+    this.socket?.disconnect();
+    this.socket?.removeAllListeners();
+    this.socket = undefined;
+  }
+
+  public async deploy(
+    repeaterId?: string
+  ): Promise<RepeaterServerDeployedEvent> {
+    if (!this.socket) {
+      throw new Error(
+        'Please make sure that repeater established a connection with host.'
+      );
+    }
+
+    this.socket.emit('deploy', {
+      repeaterId
+    });
+
+    const [result]: RepeaterServerDeployedEvent[] = await once(
+      this.socket,
+      'deployed'
+    );
+
+    return result;
+  }
+
+  public connect(hostname: string) {
+    this.socket = io(this.options.uri, {
+      parser,
+      path: '/api/ws/v1',
+      transports: ['websocket'],
+      reconnectionAttempts: this.DEFAULT_RECONNECT_TIMES,
+      auth: {
+        token: this.options.token,
+        domain: hostname
+      }
+    });
+
+    this.socket.on('connect_error', (error: Error) => {
+      logger.debug(`Unable to connect to the %s host`, this.options.uri, error);
+    });
+
+    this.createPingTimer();
+
+    logger.debug('Event bus connected to %s', this.options.uri);
+  }
+
+  public on(
+    event: RepeaterServerEvents,
+    handler: RepeaterServerEventHandlers
+  ): void {
+    if (event === 'reconnection_failed') {
+      this.onReconnectionFailed(
+        handler as RepeaterServerEventHandler<RepeaterServerReconnectionFailedEvent>
+      );
+    } else {
+      this.socket.on(event, (payload, callback) => {
```

Comment:
The socket may be undefined. https://github.com/NeuraLegion/bright-cli/pull/404/files#r1274890698